real_time/add-custom-dynamic-entity-data-source/src/main/java/com/esri/samples/add_custom_dynamic_entity_data_source/SimulatedDataSource.java

```diff
@@ -0,0 +1,180 @@
+ * the License.
+ */
+
+package com.esri.samples.add_custom_dynamic_entity_data_source;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.gson.Gson;
+
+import com.esri.arcgisruntime.data.Field;
+import com.esri.arcgisruntime.geometry.Point;
+import com.esri.arcgisruntime.geometry.SpatialReferences;
+import com.esri.arcgisruntime.realtime.DynamicEntityDataSource;
+import com.esri.arcgisruntime.realtime.DynamicEntityDataSourceInfo;
+
+/**
+ * A custom DynamicEntityDataSource for processing observations read from a given JSON file.
+ */
+class SimulatedDataSource extends DynamicEntityDataSource {
+
+  private final String fileName;
+  private final String entityIdFieldName;
+  private final long delay;
+  private final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+  private ScheduledFuture<?> observationProcessing;
+  private final AtomicInteger currentLineNumber = new AtomicInteger();
+  private final List<String> allJsonLines = new ArrayList<>();
+  private final Gson gson = new Gson();
+
+  /**
+   * Construct a custom DynamicEntityDataSource.
+   *
+   * @param fileName name of a file with JSON observation data
+   * @param entityIdFieldName the name of the field containing values that uniquely identifies each entity
+   * @param delay millisecond delay between observations being added to the data source
+   */
+  SimulatedDataSource(String fileName, String entityIdFieldName, long delay) {
+    this.fileName = fileName;
+    this.entityIdFieldName = entityIdFieldName;
+    this.delay = delay;
+  }
+
+  @Override
+  protected CompletableFuture<Void> onConnectAsync() {
+    return CompletableFuture.completedFuture((Void)null).whenComplete((unused, throwable) -> {
+      if (throwable != null) {
+        System.out.println("Failed to connect custom data source: " + throwable.getMessage());
+      } else {
+        startProcessingObservations();
+      }
+    });
+  }
+
+  /**
+   * Read a file of observation data and start processing.
+   */
+  private void startProcessingObservations() {
+    // store all lines from file in a list
+    try (BufferedReader bufferedReader = new BufferedReader(new FileReader(fileName))) {
+      allJsonLines.addAll(bufferedReader.lines().filter(Objects::nonNull).toList());
+    } catch (IOException e) {
+      System.out.println("Failed to start processing observations: " + e.getMessage());
+      throw new CompletionException(e);
```

Comment:
It is worth checking that this exception actually comes out and is not swallowed in the `whenComplete`. The doc makes it look like you might need to be looking at the stage returned by `whenComplete` to see the exception.
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletableFuture.html#whenComplete(java.util.function.BiConsumer)
>Unlike method [handle](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletionStage.html#handle(java.util.function.BiFunction)), this method is not designed to translate completion outcomes, so the supplied action should not throw an exception. However, if it does, the following rules apply: if this stage completed normally but the supplied action throws an exception, then the returned stage completes exceptionally with the supplied action's exception. Or, if this stage completed exceptionally and the supplied action throws an exception, then the returned stage completes exceptionally with this stage's exception.

Reply:
I've updated this and I've checked the error causes a retry in DynamicEntityDataSource. Not sure if that is enough.